### PR TITLE
Add stub files for pymongo

### PR DESCRIPTION
**Pull request by GrierPhillips**
This pull request is in reference to the jira ticket [Stub Files for PyMongo](https://jira.mongodb.org/browse/PYTHON-1439?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=1751108#comment-1751108).

These are the stub files that I use for pymongo version 3.5.1 and I have yet to run into any issues. That is not to say that they have been fully tested. These files would benefit from a review by someone with a better understanding of the internal operations of PyMongo as there are numerous cases where I used Any when I was unsure of an argument or return type.

### CODE REVIEW

File: pymongo/auth.pyi [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    mechanism: str
+    source: str
+    username: str
+    password: str
+    props: Any
+class GSSAPIProperties(NamedTuple):
+    service_name: str
+    canonicalize_host_name: bool
+    service_realm: Any
+def _build_credentials_tuple(mech: str, source: str, user: str, passwd: str, extra: Dict[str, Any]) -> MongoCredential: ...
```
Comment (ShaneHarvey) [inline pymongo/auth.pyi]: I don't think we should bother adding type hints to private functions/methods like this. Let's stick to things that are [publicly documented](https://github.com/mongodb/mongo-python-driver/tree/master/doc/api/pymongo).

File: pymongo/__init__.pyi [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+from pymongo.mongo_replica_set_client import MongoReplicaSetClient
+from pymongo.operations import (IndexModel,
+                                InsertOne,
+                                DeleteOne,
+                                DeleteMany,
+                                UpdateOne,
+                                UpdateMany,
+                                ReplaceOne)
+from pymongo.read_preferences import ReadPreference
+from pymongo.write_concern import WriteConcern
```
Comment (ShaneHarvey) [inline pymongo/__init__.pyi]: Are these imports necessary?

**Comment by behackett** [comment]
An emphatic yes. That said, we don't currently have the work scheduled, but everyone on the Python team here is interested.

**Comment by jessekempf** [comment]
How far away is what's on this PR from what would be acceptable for merging?

**Comment by behackett** [comment]
Honestly, I don't remember why I closed this. It might have been a mistake. We opened https://jira.mongodb.org/browse/PYTHON-1439 to track it, and it's still open.

**Comment by pcorpet** [comment]
@behackett please reopen it then :-)

**Comment by behackett** [comment]
I think I understand what happened here. The performance branch was deleted (presumably by me not realizing this would happen), which auto closed this PR. Since the target branch was deleted this can't be reopened. It would have to be recreated, preferable against master.

**Comment by MattF-NSIDC** [comment]
@behackett Thanks for your work on this PR! Is there anything I can do to help get this merged? We're excited to use official type stubs for mongo instead of using our own.

**Comment by ShaneHarvey** [comment]
I'd like to provide an update as this project was just completed. While working on this project our team realized that we would be more comfortable following a more conservative approach to adding type hints. Instead of including type annotations or stub files directly in pymongo we decided to ship stub files as a separate package:  [pymongo-stubs](https://pypi.org/project/pymongo-stubs/).

pymongo-stubs' source code is hosted on GitHub ([mongodb-labs/pymongo-stubs](https://github.com/mongodb-labs/pymongo-stubs)). The initial release is compatible with PyMongo >=3.11,<4.0 and Python >=3.6. @GrierPhillips you'll notice that you are already a contributor because I merged these changes as a starting point for the new project. Thank you for your contribution! 

Our longer term plan is to get feedback on the type hints and later (in PyMongo 4.0) incorporate them directly into the codebase as inline type annotations (in [PYTHON-2432](https://jira.mongodb.org/browse/PYTHON-2432)).

For now you are welcome to try out the stubs like this:
```
python3 -m pip install pymongo-stubs
python3 -m pip install mypy
mypy <codebase that uses pymongo>
```

Or simply install pymongo-stubs to get a better IDE experience:

**Pull request closed without merging**